Secure your brand

dzver opened the pull request:
#### Changes proposed in this Pull Request

This change adds the Secure your brand campaign

#### Testing instructions

0. Apply D49664-code
1. The country_code cookie needs to be set to 'US'
2. Keep in mind that the diff above would mean that the Domain Sandbox would operate with .com, .net, .org, and .id TLDs. The Domain Sandbox would be used if the Store Sandbox is enabled. Searching for something.com would properly suggest .com, .org, .co domains, and when the next step is clicked, the .co will be removed from the cart because it's unsupported TLD.

That's why, when searching for a domain, search a .live domain, .UK, .co.uk, .id or just deal with it. Finishing the checkout with the store sandbox would require that the domain searched is not .com, .net, or .org

##### For the `test` group
1. Sign up with a custom domain
2. Make sure the Secure your brand step is visible
<img width="1369" alt="Screenshot 2020-10-14 at 16 09 01" src="https://user-images.githubusercontent.com/82778/95993662-0d5ba780-0e38-11eb-98d0-9d081a7dff39.png">

3. You'll need to sign up twice - once by accepting the offer
<img width="1369" alt="Screenshot 2020-10-14 at 16 07 34" src="https://user-images.githubusercontent.com/82778/95993678-12b8f200-0e38-11eb-8c14-a08506bd44df.png">

4. And once by skipping it. In case the offer is skipped, the cart should only have a domain + plan
5. Try the free domain/plan - there should be no secure your brand upsell.
6. Edge case - it's possible that the domain searched has no suggested domains available. The only way to actually force that error would be to hardcode an error response in the endpoint. It should skip the step once the endpoint returns error.

##### For the `control` group
1. Make sure that signing up with a domain+plan does not render a new "Secure your brand" step

Fixes #

dzver committed:
Add the basic step, following the example from Plans

This is an initial commit

dzver committed:
Data layer for Secure your brand

dzver committed:
Use the user-provided domain for secure your brand

dzver committed:
Remove gutenboarding header

dzver committed:
Display domains

This is an optimistic display and assumes there will be direct matches.

dzver committed:
Add click handlers for going to the next step

They wouldn't add anything to cart yet

dzver committed:
Update the data layer

I used the plans as a foundation and replaced it with secure-your-brand

This change needs to be squashed

dzver committed:
WIP: add items to cart

dzver committed:
Add siteSlug as a dependency

This was preventing add to cart from sending the right request

dzver committed:
Make the skip button work

The cartItems dependency needs to be provided

dzver committed:
Skip the secure your brand step if free domain selected

dzver committed:
Make the back button work

The custom TBD thing is unnecessary, it's needed for the plans step but
not here. Secure your brand doesn't need a site, just a domain

dzver committed:
Add total cost and discount calculation

dzver committed:
Add copy and strikethrough pricing for the bundle

dzver committed:
Style it to look more like Figma

The default Apple styles do not look like the figma design. This will
have to be consulted with the designer

dzver committed:
Add painted .tlds

I'll unhardcode the color a bit later when I find the right color for it

dzver committed:
Use the appropriate color variable for TLDs

dzver committed:
Query the new endpoint in parallel with the availability check

dzver committed:
Update mobile style - expand the buttons

dzver committed:
Add a/b test config

This also adds a new flos. There are pros and cons for having a new and
it is possible to test within the onboarding flow.

dzver committed:
Fix typo

dzver committed:
Skip the secure your brand step on error

Errors can occur due to:
- A very rare domain is searched, something like google.life and all of
hte suggestions are unavailable
- There's an error with the API

dzver committed:
Missed this function when resolving merge conflicts

dzver committed:
Update the schema to reflect the most recent endpoint changes

dzver committed:
Add selector test and fixture data

dzver committed:
Unbreak adding to cart - `domains` has been renamed

dzver committed:
Linting: update includes in the files

wp-desktop reviewed (dismissed):
WordPress Desktop CI Failure for job "wp-desktop-mac".

@dzver please inspect this job's build steps for breaking changes at [this link](https://circleci.com/gh/Automattic/wp-calypso/922001). For temporal failures, you may try to "Rerun Workflow from Failed".

Please also ensure this branch is rebased off latest Calypso.

dzver committed:
Fix race condition between 2 endpoints

We don't really need a separate POST cart call to add the signup bundle,
we don't really need 2 calls, let alone 3 calls.

dzver committed:
Unbreak the skip step button

dzver committed:
Unbreak free site creation

dzver committed:
Rename action names to be closer to the other action names

dzver committed:
Use more standard names for the reducers

The decision to use plural for errors was not justified, there's no
multiple errors

Also, use the more standard items so that we don't have
secureYourBrand.secureYourBrand in the store

dzver committed:
Add reducer test

This change also updates the reducer and selectors to make the test pass

dzver committed:
Add actions test

dzver committed:
Add test for isSecureYourBrandFulfilled

dzver committed:
Fix linting problem - relative import

wp-desktop reviewed (dismissed):
WordPress Desktop CI Failure for job "wp-desktop-mac".

@dzver please inspect this job's build steps for breaking changes at [this link](https://circleci.com/gh/Automattic/wp-calypso/927180). For temporal failures, you may try to "Rerun Workflow from Failed".

Please also ensure this branch is rebased off latest Calypso.

niranjan-uma-shankar commented on client/lib/abtest/active-tests.js:
```diff
@@ -228,4 +228,13 @@ export default {
+			control: 50,
+		},
+		defaultVariation: 'control',
+		allowExistingUsers: true,
```
I think we can set this to false since this test is for `EN` only, and logged in users are going to be redirected to the new onboarding flow.

niranjan-uma-shankar commented on client/signup/steps/secure-your-brand/index.jsx:
```diff
@@ -0,0 +1,220 @@
+					</CardHeading>
+					<div className="secure-your-brand__benefits">
+						<ul>
+							<li>{ translate( 'Prevent anyone from registering (and misusing) your name' ) }</li>
```
Since this is an `en` only A/B test with unknown outcome, we don't need to submit strings for translation. That could save a few $$ and human effort :). 
We could remove the  `translate()` calls, but since that would be a bit of effort, it's also fine to keep them and  after deploying the PR, we coordinate with Team Global to deprioritize these strings so that they don't get submitted into the weekly kit.

niranjan-uma-shankar reviewed (commented):
1. This is probably an edge case, but if I do the following sequence of steps, I don't see the upsell:
select a free *.wordpress.com subdomain -> go back from the plans step -> select a custom domain -> No upsell offer seen, even though I'm assigned to the variant.

2. UI issue with really long domain names. Here's the [desktop version](https://d.pr/i/JJKbxN) and the [mobile version](https://d.pr/i/4vNVvv).

3. If I go to Map or Transfer domain, then the secure your brand step appears empty:

<img width="728" alt="Screenshot 2020-10-19 at 6 10 46 PM" src="https://user-images.githubusercontent.com/1269602/96452326-8131f180-1236-11eb-8dde-a0e048458714.png">

dzver committed:
Skip Secure your brand for mapping and transfers

dzver committed:
Rename variable as requested

dzver committed:
A test is failing

null would not be a domain registration

dzver commented on client/lib/signup/step-actions/index.js:
```diff
@@ -355,9 +362,11 @@ export function addPlanToCart( callback, dependencies, stepProvidedItems, reduxS
 	const providedDependencies = { cartItem };
 
-	const newCartItems = [ cartItem ].filter( ( item ) => item );
+	const newCartItems = [ cartItem ].filter( ( item ) => item ).concat( cartItems );
```
We discussed this topic on Slack. My arguments are that when we make multiple simultaneous API calls to the server, we can have an out-of-order `update_blog_option` that would break the signup flow. That's why the original approach to use individual API request functions didn't work.

A potential refactor here would submit the cart items at the end of the flow, and not at the end of each step.

dzver committed:
This is an attempt to resolve the back issue

It doesn't yet work, committing so that I can show it to Niranjan

dzver committed:
Make long domains be displayed with word-wrap instead of overflow

wp-desktop reviewed (dismissed):
WordPress Desktop CI Failure for job "wp-desktop-mac".

@dzver please inspect this job's build steps for breaking changes at [this link](https://circleci.com/gh/Automattic/wp-calypso/929673). For temporal failures, you may try to "Rerun Workflow from Failed".

Please also ensure this branch is rebased off latest Calypso.

dzver commented on client/signup/config/steps-pure.js:
```diff
@@ -324,6 +325,13 @@ export function generateSteps( {
+		'secure-your-brand': {
+			stepName: 'secure-your-brand',
+			dependencies: [ 'domainItem', 'siteSlug' ],
+			providesDependencies: [ 'cartItems' ],
```
I renamed `cartItems` as to `domainUpsellItems`

dzver committed:
Make the test new users only

niranjan-uma-shankar committed:
Fix: Domain bundle upsell not shown bug (#46655)

* WIP

* Switch flow and assign to test if a custom domain is selected

* Restrict to EN-US

* Not using coutnry code from state, so can be removed

wp-desktop reviewed (dismissed):
WordPress Desktop CI Failure for job "wp-desktop-mac".

@dzver please inspect this job's build steps for breaking changes at [this link](https://circleci.com/gh/Automattic/wp-calypso/931160). For temporal failures, you may try to "Rerun Workflow from Failed".

Please also ensure this branch is rebased off latest Calypso.

dzver committed:
Missing function as a result of a rebase

This rebase had too many merge conflicts, I prefer committing the
missing piece instead of fixing it in other ways

niranjan-uma-shankar reviewed (approved):
Tested scenarios, works well for me. LGTM once the PR https://github.com/Automattic/wp-calypso/pull/46719 is merged into this branch. 🚢 

An observation that I'd like to bring up, since I'm not sure if it needs to be fixed. After adding the domain upsell bundles to cart, in the checkot page if you edit cart and remove the plan item, the four domain items are still in cart with the discount applied to all four domains. Is this correct? Check screenshot below of a cart with no plan item, and discount applied to all four domain items.

<img width="914" alt="Screenshot 2020-10-23 at 5 46 35 PM" src="https://user-images.githubusercontent.com/1269602/97009581-e3099880-1561-11eb-92bc-8dc72e90e065.png">

niranjan-uma-shankar commented:
There are some failing tests, it doesn't look related to the PR but safer to re-run and confirm

dzver commented:
> An observation that I'd like to bring up, since I'm not sure if it needs to be fixed. After adding the domain upsell bundles to cart, in the checkot page if you edit cart and remove the plan item, the four domain items are still in cart with the discount applied to all four domains. Is this correct? Check screenshot below of a cart with no plan item, and discount applied to all four domain items.
> 

Yes, this is how the discount is implemented. It is possible to restrict it further but given our plans to expand on this, I think it's fine to be as is.

dzver committed:
Update the abtest datestamp to today

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/5044536

Thank you @dzver for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/5044536).
